fms/models/llava_next.py

```diff
@@ -0,0 +1,586 @@
+        self.act = str_to_activation(config.projector_hidden_act)
+        self.linear_2 = nn.Linear(
+            config.text_config.emb_dim,
+            config.text_config.emb_dim,
+            bias=config.multimodal_projector_bias,
+        )
+        self.config = config
+
+    # NOTE: HF doesn't do weight initialization for LlavaNextMultiModalProjector
+    def reset_parameters(self):
+        nn.init.xavier_uniform_(self.linear_1.weight)
+        nn.init.xavier_uniform_(self.linear_2.weight)
+        if self.config.multimodal_projector_bias:
+            nn.init.normal_(self.linear_1.bias, std=1e-6)
+            nn.init.normal_(self.linear_2.bias, std=1e-6)
+
+    def forward(self, image_features):
+        hidden_states = self.linear_1(image_features)
+        hidden_states = self.act(hidden_states)
+        hidden_states = self.linear_2(hidden_states)
+        return hidden_states
+
+
+class LlavaNext(nn.Module):
+    def __init__(
+        self,
+        config: Optional[LlavaNextConfig] = None,
+        distributed_strategy: DistributedStrategy = NoOpStrategy,
+        **kwargs,
+    ):
+        super(LlavaNext, self).__init__()
+
+        if config is not None:
+            self.config = config
+        else:
+            self.config = LlavaNextConfig()
+        self.config = self.config.updated(**kwargs)
+        self.distributed_strategy = distributed_strategy
+
+        if not isinstance(self.config.vision_config, SiglipVisionConfig):
+            print(
+                "FMS implementation of LlavaNext currently supports only Siglip vision model"
+            )
+        if not isinstance(self.config.text_config, GraniteConfig):
+            print(
+                "FMS implementation of LlavaNext currently supports only Granite language model"
+            )
+
+        # Only supporting granite text decoder encoder for now
+        self.language_model = Granite(self.config.text_config)
+
+        # Only supporting siglip vision encoder for now
+        self.vision_tower = SiglipVision(self.config.vision_config)
+
+        self.multi_modal_projector = LlavaNextMultiModalProjector(self.config)
+        embed_std = 1 / math.sqrt(self.config.text_config.emb_dim)
+        self.image_newline = nn.Parameter(
+            torch.randn(self.config.text_config.emb_dim) * embed_std
+        )
+        self.vocab_size = self.config.text_config.src_vocab_size
+
+    @classmethod
+    def from_config(cls, config: LlavaNextConfig) -> "LlavaNext":
+        return cls(config)
+
+    def get_config(self) -> LlavaNextConfig:
+        return self.config
+
+    def reset_parameters(self):
+        nn.init.xavier_uniform_(self.image_newline.data)
+        self.langauage_model.reset_parameters()
+        self.vision_tower.reset_parameters()
+        self.multi_modal_projector.reset_parameters()
+
+    def post_init(self):
+        self.language_model.post_init()
+        self.vision_tower.post_init()
+
+    def unpad_image(self, tensor, original_size):
```

Comment:
in this function and the next few ones, can you add types for mypy? This will also help me follow the code and understand better why there is a graph break happening

Reply:
Will add